File: src/main/java/com/github/lizhanyin/tfs/wizard/step/ProjectSelectionStep.java
package com.github.lizhanyin.tfs.wizard.step;

import com.github.lizhanyin.tfs.services.TfsConnectionService;
import com.github.lizhanyin.tfs.wizard.ImportProjectContext;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import com.intellij.ui.CheckboxTree;
import com.intellij.ui.CheckedTreeNode;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.util.ui.FormBuilder;
import com.intellij.util.ui.JBUI;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.tree.DefaultTreeModel;
import javax.swing.tree.TreeNode;
import java.awt.*;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;

/**
 * 项目选择步骤
 */
public class ProjectSelectionStep extends AbstractWizardStep {

    private static final String STEP_ID = "project-selection";

    // UI 组件
    private CheckboxTree projectTree;
    private CheckedTreeNode rootTreeNode;
    private TextFieldWithBrowseButton localPathField;
    private JBLabel statusLabel;
    private JButton refreshButton;

    // 项目列表
    private final List<ProjectItem> projects = new ArrayList<>();

    public ProjectSelectionStep(@NotNull ImportProjectContext context) {
        super(STEP_ID, "选择项目", context);
    }

    @Override
    protected JComponent buildComponent() {
        FormBuilder builder = FormBuilder.createFormBuilder();

        // 本地路径选择
        localPathField = new TextFieldWithBrowseButton();
        localPathField.getTextField().setText(context.getLocalPath() != null ? context.getLocalPath() : System.getProperty("user.home"));
        localPathField.addBrowseFolderListener(
                null,
                FileChooserDescriptorFactory.createSingleFolderDescriptor()
        );
        builder.addLabeledComponent("本地路径:", localPathField);

        builder.addSeparator();

        // 项目树标签
        JPanel headerPanel = new JPanel(new BorderLayout());
        headerPanel.add(new JBLabel("选择要导入的项目/文件夹:"), BorderLayout.WEST);

        refreshButton = new JButton("刷新");
        refreshButton.addActionListener(e -> loadProjects());
        headerPanel.add(refreshButton, BorderLayout.EAST);

        builder.addComponent(headerPanel);

        // 创建项目树
        createProjectTree();
        JBScrollPane treeScrollPane = new JBScrollPane(projectTree);
        treeScrollPane.setPreferredSize(new Dimension(400, 300));
        builder.addComponentFillVertically(treeScrollPane, 0);

        builder.addSeparator();

        // 状态标签
        statusLabel = new JBLabel("正在加载项目列表...");
        builder.addComponent(statusLabel);

        JPanel panel = builder.getPanel();
        panel.setBorder(JBUI.Borders.empty(10));

        // 初始加载
        loadProjects();

        return panel;
    }

    /**
     * 创建项目树
     */
    private void createProjectTree() {
        rootTreeNode = new CheckedTreeNode(null);

        projectTree = new CheckboxTree(new CheckboxTree.CheckboxTreeCellRenderer() {
            public void customizeCellRenderer(@NotNull JTree tree, Object value, boolean selected, boolean expanded, boolean leaf, int row, boolean hasFocus) {
                if (value instanceof CheckedTreeNode) {
                    Object userObject = ((CheckedTreeNode) value).getUserObject();
                    if (userObject instanceof ProjectItem) {
                        getTextRenderer().append(((ProjectItem) userObject).getName());
                    }
                }
            }
        }, rootTreeNode);
        projectTree.setRootVisible(false);
        projectTree.setShowsRootHandles(true);
    }

    /**
     * 加载项目列表
     */
    private void loadProjects() {
        if (!context.isTeamProjectSelected()) {
            statusLabel.setText("请先选择团队项目");
            statusLabel.setForeground(java.awt.Color.RED);
            return;
        }

        statusLabel.setText("正在加载项目列表...");
        statusLabel.setForeground(java.awt.Color.BLACK);
        refreshButton.setEnabled(false);

        ProgressManager.getInstance().run(new Task.Backgroundable(null, "加载项目", false) {
            @Override
            public void run(@NotNull ProgressIndicator indicator) {
                indicator.setText("正在从 TFS 服务器获取项目列表...");
                indicator.setIndeterminate(true);

                try {
                    TfsConnectionService connectionService =
                            ApplicationManager.getApplication().getService(TfsConnectionService.class);

                    List<TfsConnectionService.ProjectItemInfo> serverItems =
                            connectionService.getProjectItems(context, context.getTeamProject());

                    // 转换为本地项目项
                    List<ProjectItem> projectItems = new ArrayList<>();
                    for (TfsConnectionService.ProjectItemInfo item : serverItems) {
                        if (item.isFolder()) {
                            ProjectItem projectItem = new ProjectItem(item.getServerPath(), item.getName());
                            // 如果是文件夹，递归加载子项
                            loadChildren(projectItem, connectionService, indicator);
                            projectItems.add(projectItem);
                        }
                    }

                    // 更新 UI
                    SwingUtilities.invokeLater(() -> {
                        projects.clear();
                        projects.addAll(projectItems);
                        buildProjectTree();

                        int itemCount = countLeafNodes(rootTreeNode);
                        if (itemCount == 0) {
                            statusLabel.setText("未找到可导入的项目");
                            statusLabel.setForeground(java.awt.Color.ORANGE);
                        } else {
                            statusLabel.setText("已加载 " + itemCount + " 个项目/文件夹");
                            statusLabel.setForeground(new java.awt.Color(0, 128, 0));
                        }
                        refreshButton.setEnabled(true);
                    });

                } catch (Exception e) {
                    SwingUtilities.invokeLater(() -> {
                        statusLabel.setText("加载失败: " + e.getMessage());
                        statusLabel.setForeground(java.awt.Color.RED);
                        refreshButton.setEnabled(true);
                    });
                }
            }
        });
    }

    /**
     * 递归加载子项
     */
    private void loadChildren(ProjectItem parent, TfsConnectionService connectionService, ProgressIndicator indicator) {
        try {
            List<TfsConnectionService.ProjectItemInfo> children =
                    connectionService.getChildItems(context, parent.getServerPath());

            for (TfsConnectionService.ProjectItemInfo child : children) {
                if (child.isFolder()) {
                    ProjectItem childItem = new ProjectItem(child.getServerPath(), child.getName());
                    parent.addChild(childItem);
                    // 限制递归深度
                    if (parent.getServerPath().split("/").length < 6) {
                        loadChildren(childItem, connectionService, indicator);
                    }
                }
            }
        } catch (Exception e) {
            // 忽略单个文件夹加载失败
        }
    }

    /**
     * 构建项目树
     */
    private void buildProjectTree() {
        rootTreeNode.removeAllChildren();

        for (ProjectItem item : projects) {
            CheckedTreeNode node = createTreeNode(item);
            rootTreeNode.add(node);
        }

        ((DefaultTreeModel) projectTree.getModel()).reload();
    }

    /**
     * 创建树节点
     */
    private CheckedTreeNode createTreeNode(ProjectItem item) {
        CheckedTreeNode node = new CheckedTreeNode(item);

        if (item.getChildren() != null) {
            for (ProjectItem child : item.getChildren()) {
                node.add(createTreeNode(child));
            }
        }

        return node;
    }

    /**
     * 统计叶子节点数量
     */
    private int countLeafNodes(TreeNode node) {
        if (node.isLeaf()) {
            return 1;
        }

        int count = 0;
        Enumeration<TreeNode> children = (Enumeration<TreeNode>) node.children();
        while (children.hasMoreElements()) {
            count += countLeafNodes(children.nextElement());
        }
        return count;
    }

    /**
     * 获取选中的项目路径
     */
    private List<String> getSelectedPaths() {
        List<String> paths = new ArrayList<>();
        collectCheckedPaths(rootTreeNode, paths);
        return paths;
    }

    /**
     * 递归收集选中的路径
     */
    private void collectCheckedPaths(CheckedTreeNode node, List<String> paths) {
        if (node.isLeaf() && node.isChecked()) {
            Object userObject = node.getUserObject();
            if (userObject instanceof ProjectItem) {
                paths.add(((ProjectItem) userObject).getServerPath());
            }
        }

        Enumeration<TreeNode> children = node.children();
        while (children.hasMoreElements()) {
            TreeNode child = children.nextElement();
            if (child instanceof CheckedTreeNode) {
                collectCheckedPaths((CheckedTreeNode) child, paths);
            }
        }
    }

    @Override
    public boolean isComplete() {
        List<String> selected = getSelectedPaths();
        return !selected.isEmpty() && !localPathField.getText().isEmpty();
    }

    @Override
    public JComponent getPreferredFocusedComponent() {
        return projectTree;
    }

    @Override
    public boolean onFinish() {
        List<String> selectedPaths = getSelectedPaths();
        context.setSelectedProjects(selectedPaths);
        context.setLocalPath(localPathField.getText());
        return true;
    }

    /**
     * 项目项
     */
    private static class ProjectItem {
        private final String serverPath;
        private final String name;
        private List<ProjectItem> children;

        public ProjectItem(String serverPath, String name) {
            this.serverPath = serverPath;
            this.name = name;
        }

        public String getServerPath() { return serverPath; }
        public String getName() { return name; }
        public List<ProjectItem> getChildren() { return children; }

        public void addChild(ProjectItem child) {
            if (children == null) {
                children = new ArrayList<>();
            }
            children.add(child);
        }

        @Override
        public String toString() {
            return name;
        }
    }
}

File: src/main/kotlin/com/github/lizhanyin/tfs/wizard/step/ProjectSelectionStep.kt
package com.github.lizhanyin.tfs.wizard.step

import com.github.lizhanyin.tfs.TfsBundle
import com.github.lizhanyin.tfs.client.codemarker.CodeMarker
import com.github.lizhanyin.tfs.client.codemarker.CodeMarkerDispatch
import com.github.lizhanyin.tfs.client.ui.vc.serveritem.TypedServerItem
import com.github.lizhanyin.tfs.services.TfsConnectionService
import com.github.lizhanyin.tfs.vc.serveritem.ServerItemLabelProvider
import com.github.lizhanyin.tfs.wizard.ImportProjectContext
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.icons.AllIcons
import com.intellij.openapi.ui.Messages
import com.intellij.ui.JBColor
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.treeStructure.Tree
import com.intellij.util.ui.JBUI
import java.awt.BorderLayout
import java.awt.Dimension
import javax.swing.JButton
import javax.swing.JComponent
import javax.swing.JPanel
import javax.swing.SwingUtilities
import javax.swing.event.TreeSelectionListener
import javax.swing.tree.DefaultMutableTreeNode
import javax.swing.tree.DefaultTreeModel
import javax.swing.tree.TreeSelectionModel

/**
 * 项目选择步骤（第四步）
 *
 * UI 布局（参考 CLAUDE.md）：
 * ------------------------------------------
 * |  项目选择                                |
 * |    从 Team Foundation Server 导入项目    |
 * |----------------------------------------|
 * | 导入项目:                                |
 * | [tree] (ip\集合名称）                    |
 * |    root  (项目名）                       |
 * |      文件夹1                            |
 * |      文件夹2                            |
 * | [label] (已选择 {0} 个项目)              |
 * | [checkbox] 对选定项目中的文件执行强制获取  |
 * |----------------------------------------|
 * | [上一步] [下一步]  [完成](禁用) [放弃]    |
 * ------------------------------------------
 */
class ProjectSelectionStep(context: ImportProjectContext) :
    AbstractWizardStep(STEP_ID, TfsBundle.message("ProjectSelectionStep.title"), context, TfsBundle.message("ProjectSelectionStep.description")) {

    companion object {
        private const val STEP_ID = "project-selection"
        private val CODEMARKER_CHILD_NODES_FETCH_START =
        CodeMarker("com.github.lizhanyin.tfs.wizard.step.ProjectSelectionStep#childNodesFetchStart"); //$NON-NLS-1$
        private val CODEMARKER_CHILD_NODES_FETCH_COMPLETE =
            CodeMarker("com.github.lizhanyin.tfs.wizard.step.ProjectSelectionStep#childNodesFetchComplete"); //$NON-NLS-1$
    }

    private lateinit var projectTree: Tree
    private lateinit var rootTreeNode: DefaultMutableTreeNode
    private lateinit var selectedCountLabel: JBLabel
    private lateinit var forceGetCheckbox: JBCheckBox
    private lateinit var statusLabel: JBLabel
    private lateinit var refreshButton: JButton

    private var labelProvider: ServerItemLabelProvider = ServerItemLabelProvider()

    private val projects = mutableListOf<ProjectItem>()

    override fun buildComponent(): JComponent {
        val contentPanel = JPanel(BorderLayout(0, JBUI.scale(8)))
        contentPanel.border = JBUI.Borders.empty(10)

        // 中间：项目树 + 状态
        contentPanel.add(createTreePanel(), BorderLayout.CENTER)

        // 底部：选择计数 + 强制获取复选框
        contentPanel.add(createBottomPanel(), BorderLayout.SOUTH)

        // 初始加载
        loadProjects()

        return contentPanel
    }

    // ==================== UI 构建 ====================

    private fun createTreePanel(): JComponent {
        val panel = JPanel(BorderLayout(0, JBUI.scale(4)))

        // 树标题 + 刷新按钮
        val headerPanel = JPanel(BorderLayout())
        headerPanel.add(JBLabel(TfsBundle.message("ProjectSelectionStep.label.selectProjects")), BorderLayout.WEST)

        refreshButton = JButton(TfsBundle.message("ProjectSelectionStep.button.refresh"))
        refreshButton.addActionListener { loadProjects() }
        headerPanel.add(refreshButton, BorderLayout.EAST)
        panel.add(headerPanel, BorderLayout.NORTH)

        // 创建项目树
        createProjectTree()
        val treeScrollPane = JBScrollPane(projectTree)
        treeScrollPane.preferredSize = Dimension(500, 300)
        panel.add(treeScrollPane, BorderLayout.CENTER)

        // 状态标签
        statusLabel = JBLabel(TfsBundle.message("ProjectSelectionStep.status.loading"))
        statusLabel.foreground = JBColor.gray
        panel.add(statusLabel, BorderLayout.SOUTH)

        return panel
    }

    private fun createBottomPanel(): JComponent {
        val panel = JPanel(BorderLayout())
        panel.border = JBUI.Borders.emptyTop(8)

        selectedCountLabel = JBLabel(TfsBundle.message("ProjectSelectionStep.status.selectedNone"))
        panel.add(selectedCountLabel, BorderLayout.WEST)

        forceGetCheckbox = JBCheckBox(TfsBundle.message("ProjectSelectionStep.checkbox.forceGet"))
        panel.add(forceGetCheckbox, BorderLayout.EAST)

        return panel
    }

    private fun createProjectTree() {
        rootTreeNode = DefaultMutableTreeNode()

        projectTree = Tree(rootTreeNode)
        projectTree.isRootVisible = false
        projectTree.setShowsRootHandles(true)
        projectTree.selectionModel.selectionMode = TreeSelectionModel.DISCONTIGUOUS_TREE_SELECTION
        projectTree.setCellRenderer { tree, value, selected, expanded, leaf, row, hasFocus ->
            val label = javax.swing.JLabel()
            if (value is DefaultMutableTreeNode) {
                val userObject = value.userObject
                if (userObject is ProjectItem) {
                    label.icon = labelProvider.getImage(userObject.node)
                    label.text = labelProvider.getText(userObject.node)
                }
            }
            label
        }

        projectTree.addTreeSelectionListener(TreeSelectionListener { updateSelectedCount() })
    }

    // ==================== 数据加载 ====================

    private fun loadProjects() {
        if (!context.isTeamProjectSelected()) {
            statusLabel.text = TfsBundle.message("ProjectSelectionStep.error.selectTeamProjectFirst")
            statusLabel.foreground = JBColor.RED
            return
        }

        statusLabel.text = TfsBundle.message("ProjectSelectionStep.status.loading")
        statusLabel.foreground = JBColor.gray
        refreshButton.isEnabled = false

        ProgressManager.getInstance().run(object : Task.Backgroundable(null, TfsBundle.message("ProjectSelectionStep.progress.loading"), false) {
            override fun run(indicator: ProgressIndicator) {
                indicator.text = TfsBundle.message("ProjectSelectionStep.progress.fetching")
                indicator.isIndeterminate = true

                try {
                    val projectItems = mutableListOf<ProjectItem>()

                    val connectionService = ApplicationManager.getApplication().getService(TfsConnectionService::class.java)
                    // 根节点 ServerItemType
                    val itemSource = connectionService.getChildItems(context)
                    labelProvider.setServerItemSource(itemSource)

                    val root = ProjectItem(TypedServerItem.ROOT, labelProvider);
                    projectItems.add(root)

                    // 子节点 TypedServerItem
                    CodeMarkerDispatch.dispatch(CODEMARKER_CHILD_NODES_FETCH_START)
                    val children = itemSource.getChildren(root.node);

                    CodeMarkerDispatch.dispatch(CODEMARKER_CHILD_NODES_FETCH_COMPLETE)

                    if (children.size == 0) {
                        return
                    }

                    loadChildren(root, indicator, root.node.serverPath)

                    SwingUtilities.invokeLater {
                        projects.clear()
                        projects.addAll(projectItems)
                        buildProjectTree()

                        if (projects.isEmpty()) {
                            statusLabel.text = TfsBundle.message("ProjectSelectionStep.status.noResults")
                            statusLabel.foreground = JBColor.ORANGE
                        } else {
                            statusLabel.text = TfsBundle.message("ProjectSelectionStep.status.loaded", projects.size)
                            statusLabel.foreground = JBColor.GRAY
                        }
                        refreshButton.isEnabled = true
                        updateSelectedCount()
                    }
                } catch (e: Exception) {
                    SwingUtilities.invokeLater {
                        Messages.showErrorDialog(
                            TfsBundle.message("ProjectSelectionStep.error.loadingFailed", e.message ?: ""),
                            TfsBundle.message("ProjectSelectionStep.title")
                        )
                        refreshButton.isEnabled = true
                    }
                }
            }
        })
    }

    private fun loadChildren(
        parent: ProjectItem,
        indicator: ProgressIndicator,
        rootPath: String
    ) {
        if (indicator.isCanceled) return

        try {
            val children = labelProvider.getServerItemSource().getChildren(parent.node)

            for (child in children) {
                val childItem = ProjectItem(child, labelProvider)
                parent.addChild(childItem)
                // 限制递归深度（相对于根路径最多 4 层）
                val depth = child.serverPath.removePrefix(rootPath).count { it == '/' }
                if (depth < 1) {
                    loadChildren(childItem, indicator, rootPath)
                }
            }
        } catch (_: Exception) {
            // 忽略单个文件夹加载失败
        }
    }

    // ==================== 树操作 ====================

    private fun buildProjectTree() {
        rootTreeNode.removeAllChildren()

        for (item in projects) {
            val node = createTreeNode(item)
            rootTreeNode.add(node)
        }

        (projectTree.model as DefaultTreeModel).reload()
    }

    private fun createTreeNode(item: ProjectItem): DefaultMutableTreeNode {
        val node = DefaultMutableTreeNode(item)

        item.children?.forEach { child ->
            node.add(createTreeNode(child))
        }

        return node
    }

    private fun updateSelectedCount() {
        val count = getSelectedPaths().size
        selectedCountLabel.text = if (count == 0) {
            TfsBundle.message("ProjectSelectionStep.status.selectedNone")
        } else {
            TfsBundle.message("ProjectSelectionStep.status.selectedCount", count)
        }
    }

    // ==================== 选择相关 ====================

    private fun getSelectedPaths(): List<String> {
        return projectTree.selectionPaths?.mapNotNull { path ->
            val node = path.lastPathComponent as? DefaultMutableTreeNode
            val item = node?.userObject as? ProjectItem
            item?.node?.serverPath
        } ?: emptyList()
    }

    // ==================== 步骤接口 ====================

    override fun isComplete(): Boolean = getSelectedPaths().isNotEmpty()

    override fun getPreferredFocusedComponent(): JComponent = projectTree

    override fun onFinish(): Boolean {
        val selectedPaths = getSelectedPaths()
        context.selectedProjects.clear()
        context.selectedProjects.addAll(selectedPaths)
        context.forceGetLatest = forceGetCheckbox.isSelected
        return true
    }

    /**
     * 项目项
     */
    private class ProjectItem(
        val node: TypedServerItem,
        val labelProvider: ServerItemLabelProvider
    ) {
        var children: MutableList<ProjectItem>? = null

        fun addChild(child: ProjectItem) {
            if (children == null) {
                children = mutableListOf()
            }
            children!!.add(child)
        }

        override fun toString(): String = labelProvider.getText(node)
    }
}
